diff -urBN sys/fs/unionfs.orig/union_subr.c sys/fs/unionfs/union_subr.c
--- ../src/sys/fs/unionfs.orig/union_subr.c	Tue Jul 11 03:05:51 2006
+++ ../src/sys/fs/unionfs/union_subr.c	Mon Jul 10 23:56:38 2006
@@ -626,11 +626,13 @@
 	 */
 	VI_LOCK(vp);
 	unp->un_uppervp = uvp;
+	vp->v_vnlock = uvp->v_vnlock;
 	lockcnt = lvp->v_vnlock->lk_exclusivecount;
+	if (0 >= lockcnt)
+		panic("unionfs: no exclusive lock");
+	VI_UNLOCK(vp);
 	for (count = 0; count < lockcnt; count++)
 		VOP_UNLOCK(lvp, 0, td);
-	vp->v_vnlock = uvp->v_vnlock;
-	VI_UNLOCK(vp);
 	for (count = 1; count < lockcnt; count++)
 		vn_lock(uvp, LK_EXCLUSIVE | LK_CANRECURSE | LK_RETRY, td);
 
diff -urBN sys/fs/unionfs.orig/union_vfsops.c sys/fs/unionfs/union_vfsops.c
--- ../src/sys/fs/unionfs.orig/union_vfsops.c	Tue Jul 11 03:06:00 2006
+++ ../src/sys/fs/unionfs/union_vfsops.c	Mon Jul 10 23:05:03 2006
@@ -143,10 +143,12 @@
 	 * Check multi unionfs mount to avoid ock against myself' panic.
 	 */
 	lowerrootvp = mp->mnt_vnodecovered;
+/*
 	if (lowerrootvp->v_mount->mnt_op == &unionfs_vfsops) {
 		UNIONFSDEBUG("unionfs_mount: multi unionfs mount?\n");
 		return (EDEADLK);
 	}
+*/
 	/*
 	 * Get argument (part 1)
 	 */
